no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralVisitor.java

```diff
@@ -0,0 +1,244 @@
+   * are found.
+   *
+   * @param atm the type to search
+   * @return true if the type or any of its components has a non-default annotation
+   */
+  private boolean hasNonDefault(final AnnotatedTypeMirror atm) {
+    NonDefaultScanner scanner = new NonDefaultScanner();
+    scanner.visit(atm);
+    return scanner.getResult();
+  }
+
+  /**
+   * This checker defaults implicit upper bounds of type variables to @NonConstant. This is the
+   * correct thing to do in unannotated code, but causes false positive errors whenever a user
+   * writes a type annotation on the upper bound of a type variable (for instance, by declaring a
+   * list of constant strings).
+   *
+   * <p>For example, without overriding this method, this code would not typecheck:
+   *
+   * <p>{@code List<@MaybeDerivedFromConstant String> list = new ArrayList<>();}
+   *
+   * <p>Since this is common, this code disables checking that type arguments supplied to a type or
+   * a method invocation are within the bounds of the type variables as declared for user-written
+   * annotations. Stolen shamelessly from https://github.com/awslabs/data-classification-checker
+   */
+  @Override
+  protected void checkTypeArguments(
+      final Tree toptree,
+      final List<? extends AnnotatedTypeParameterBounds> paramBounds,
+      final List<? extends AnnotatedTypeMirror> typeargs,
+      final List<? extends Tree> typeargTrees) {
+    List<AnnotatedTypeParameterBounds> newParamBounds = new ArrayList<>();
+    List<AnnotatedTypeMirror> newTypeArgs = new ArrayList<>();
+    for (int i = 0; i < typeargs.size(); i++) {
+      AnnotatedTypeMirror atm = typeargs.get(i);
+      if (!hasNonDefault(atm)) {
+        newParamBounds.add(paramBounds.get(i));
+        newTypeArgs.add(atm);
+      }
+    }
+    if (!newTypeArgs.isEmpty()) {
+      super.checkTypeArguments(toptree, newParamBounds, newTypeArgs, typeargTrees);
+    }
+  }
+
+  /**
+   * Skip the standard subtyping check on receivers on all method calls. Constants can't have
+   * methods called on them, so it doesn't matter.
+   */
+  @Override
+  protected boolean skipReceiverSubtypeCheck(
+      final MethodInvocationTree node,
+      final AnnotatedTypeMirror methodDefinitionReceiver,
+      final AnnotatedTypeMirror methodCallReceiver) {
+    return true;
+  }
+
+  @Override
+  protected void checkConstructorResult(
+      AnnotatedExecutableType constructorType, ExecutableElement constructorElement) {
+    // Skip this and do nothing. Normally, this method will issue a warning when
+    // the type of a class defaults to something other than top. That's fine
+    // for this checker, because we don't expect user-written classes to take
+    // on values other than @NonConstant.
+  }
+
+  /**
+   * The standard override checker is replaced so that no override.param.invalid errors are issued
+   * when overriding methods defined in bytecode (which therefore use a different, incompatible
```

Comment:
See my comment above on the code that actually implements it - I think it's sound, because it basically pretends for the purpose of override checking that bytecode uses conservative defaults.